buildchain/dodo.py

```diff
@@ -23,29 +25,92 @@
+
+    def __init__(self, outstream: TextIO, options: Dict[str, Any]):
+        super().__init__(outstream, options)
+        self.failures : List[Dict[str, Any]] = []
+        self.runtime_errors : List[str] = []
 
     def execute_task(self, task: doit.task.Task) -> None:
         """Called when a task is executed."""
-        # Ignore tasks that do not define actions.
-        if task.actions:
-            self.write('.  {}\n'.format(task.title()))
+        super().execute_task(task)
+        if task.actions:  # Ignore tasks that do not define actions.
+            self._write('.  {}\n'.format(task.title()))
+
+    def add_failure(
+        self, task: doit.task.Task, exception: doit.exceptions.CatchedException
+    ) -> None:
+        """Called when execution finishes with a failure"""
+        super().add_failure(task, exception)
+        result = {'task': task, 'exception':exception}
+        self.failures.append(result)
+        self._write_failure(result)
 
     def skip_uptodate(self, task: doit.task.Task) -> None:
         """Called when a task is skipped (up-to-date)."""
-        self.write('-- {}\n'.format(task.title()))
+        super().skip_uptodate(task)
+        self._write('-- {}\n'.format(task.title()))
 
     def skip_ignore(self, task: doit.task.Task) -> None:
         """Called when a task is skipped (ignored)."""
-        self.write('!! {}\n'.format(task.title()))
+        super().skip_ignore(task)
+        self._write('!! {}\n'.format(task.title()))
+
+    def cleanup_error(
+        self, exception: doit.exceptions.CatchedException
+    ) -> None:
+        """Error during cleanup."""
+        self._write(exception.get_msg())
+
+    def runtime_error(self, msg: str) -> None:
+        """Error from doit itself (not from a task execution)."""
+        # saved so they are displayed after task failures messages
+        self.runtime_errors.append(msg)
+
+    def complete_run(self) -> None:
+        """Called when finished running all tasks."""
+        # If test fails print output from failed task.
+        for result in self.failures:
+            task = result['task']
+            # Makes no sense to print output if task was not executed.
+            if not task.executed:
+                continue
+            self._write('#'*40 + '\n')
+            self._write_failure(result)
+            err = ''.join([action.err for action in task.actions if action.err])
+            self._write('{} <stderr>:\n{}\n'.format(task.name, err))
+            out = ''.join([action.out for action in task.actions if action.out])
+            self._write('{} <stdout>:\n{}\n'.format(task.name, out))
+
+        if self.runtime_errors:
+            self._write('#'*40 + '\n')
+            self._write('Execution aborted.\n')
+            self._write('\n'.join(self.runtime_errors))
+            self._write('\n')
+        # Generate the build log.
+        build_log = constants.ROOT/'build.log'
```

Comment:
This variable might be better suited higher in the hierarchy - or even in `constants` - as it's a fairly important one (also, a constant would be cool for @NicolasT 's eve-in-python PoC)

Reply:
It's only used locally, so not sure it make sense to expose it in a wider scope.

Can be moved when the need arises.

Reply:
Moving it to class variable would be better for visibility imo, in case anyone should be looking for it.

Not a big issue though.